### PR TITLE
106 media permissions

### PR DESCRIPTION
**Pull request by andreibudisan**

### CODE REVIEW

File: Source/Helpers/YPPermissionCheckable.swift [hunk 1]
```diff
@@ -30,21 +30,38 @@ extension YPPermissionCheckable where Self: UIViewController {
     // Async beacause will prompt permission if .notDetermined
     // and ask custom popup if denied.
     func checkPermissionToAccessVideo(block: @escaping (Bool) -> Void) {
-        switch AVCaptureDevice.authorizationStatus(for: AVMediaType.video) {
+        
+        switch AVCaptureDevice.authorizationStatus(for: .video) {
         case .authorized:
-            block(true)
+            
+            switch AVAudioSession.sharedInstance().recordPermission {
```
[the remark this answers is not in the document]
Comment (andreibudisan) [reply Source/Helpers/YPPermissionCheckable.swift]: When requesting permission for camera, the user is also requested microphone permission. In case the microphone permission is denied this check is made

File: Source/Models/YPPermissionDeniedPopup.swift [hunk 2]
```diff
@@ -8,28 +8,59 @@
...
+    }
+    
+    func setup(for permissionType: PermissionType) {
+        
+        self.permissionType = permissionType
+    }
+    
+    private func setupUI() {
+        
+        title = YPConfig.wordings.permissionPopup.title
```
Comment (alinmuntean) [inline Source/Models/YPPermissionDeniedPopup.swift:47]: What happens if I actually do set a `title` and a `message` when calling the constructor on this `UIAlertController` class? Will they be taken into consideration if they are different than `nil`? Just create an extension over UIAlertController which takes those 2 needed items (`cancel closure` and `PermissionType`) as parameters. It can be something like:

> extension UIAlertController {
>     class func permissionDeniedAlert(forType permissionType: PermissionType, onCancel: (() -> Void)? = nil) -> UIAlertController
> }

File: Source/Helpers/YPPermissionCheckable.swift [hunk 3]
```diff
@@ -19,33 +19,72 @@ extension YPPermissionCheckable where Self: UIViewController {
...
-            block(true)
+            completion(true)
+            
         case .restricted, .denied:
-            let popup = YPPermissionDeniedPopup()
-            let alert = popup.popup(cancelBlock: {
-                block(false)
-            })
+            let alert = UIAlertController.permissionDeniedAlert(forType: .camera) {
+                completion(false)
```
Comment (alinmuntean) [inline Source/Helpers/YPPermissionCheckable.swift:48]: It's hard to guess what's this closure for, is it for when the alert is shown/dismissed? Is it when the Settings button is pressed? Is it when the Cancel button is pressed? In those cases, when it's hard/difficult to read/understand the code, it's better to specify the name of the block (i.e. `onCancel: {`).

File: Source/Pages/Gallery/YPLibraryVC.swift [hunk 4]
```diff
@@ -228,13 +228,19 @@ public class YPLibraryVC: UIViewController, YPPermissionCheckable {
     }
     
     func checkPermission() {
+        
         checkPermissionToAccessPhotoLibrary { [weak self] hasPermission in
-            guard let strongSelf = self else {
```
Comment (alinmuntean) [inline Source/Pages/Gallery/YPLibraryVC.swift:232]: Why can't we keep the checking of valid `self` here and refer to `strongSelf` on the following lines?

File: Source/Pages/Gallery/YPLibraryVC.swift [hunk 5]
```diff
@@ -228,13 +228,19 @@ public class YPLibraryVC: UIViewController, YPPermissionCheckable {
...
-                return
-            }
-            if hasPermission && !strongSelf.initialized {
-                strongSelf.initialize()
-                strongSelf.initialized = true
+            
+            if hasPermission {
```
Comment (alinmuntean) [inline Source/Pages/Gallery/YPLibraryVC.swift:234]: Would it work this way?

> if !hasPermission {
            strongSelf.dismiss(animated: true)
          } else if !strongSelf.initialized {
            strongSelf.initialize()
          }
> 
In this case you have to move the `strongSelf.initialized = true` inside the `.initialize()` function, at the end maybe.

File: Source/Pages/Photo/YPCameraVC.swift [hunk 6]
```diff
@@ -54,26 +54,34 @@ public class YPCameraVC: UIViewController, UIGestureRecognizerDelegate, YPPermis
...
-            }
-            self?.photoCapture.start(with: strongSelf.v.previewViewContainer, completion: {
-                DispatchQueue.main.async {
-                    self?.isInited = true
-                    self?.refreshFlashButton()
+        
+        doAfterCameraPermissionCheck { [weak self] granted in
+            
+            if granted {
+                guard let strongSelf = self else {
```
Comment (alinmuntean) [inline Source/Pages/Photo/YPCameraVC.swift:61]: Same as above, try to make it easier to read.

File: Source/Pages/Video/YPVideoCaptureVC.swift [hunk 7]
```diff
@@ -55,19 +55,26 @@ public class YPVideoCaptureVC: UIViewController, YPPermissionCheckable {
...
     func start() {
+        
         v.shotButton.isEnabled = false
-        doAfterPermissionCheck { [weak self] in
-            guard let strongSelf = self else {
-                return
+        doAfterVideoPermissionCheck { [weak self] granted in
+            
+            if granted {
+                guard let strongSelf = self else {
```
Comment (alinmuntean) [inline Source/Pages/Video/YPVideoCaptureVC.swift:63]: Same as above.

**Comment by RolandTolnay** [comment]
Considering this code was not written by us, I would not invest a lot of effort in maintaining it or scaling it.  
The plan in the future is to deprecate it anyway, replacing it with our own solution.

Therefore during the review I was focused mostly on functionality and edge cases, and less on code quality.
As long as flows are covered and behave in an acceptable way, the task can be closed.

However there are still a few cases which are not properly covered, and as such should be resolved before we can close this task. 
I have described them in detail below:

### 1.
**Prerequisites**: Fresh install of the app
**Steps to reproduce**:
1. Log-in online and navigate into a folder
2. Go to Folder actions and tap "Take a picture"
3. When asked for permissions, press "Don't allow"
4. Go to Folder actions and tap "Take a picture" again
5. An alert will appear, saying you need permissions. Tap on "Settings"
6. **Do not change anything in Settings**. Go back to the app
7. Observe the black screen still there. 
8. Tap on the record button (or on the black screen)
9. The same alert will appear again, saying you need permissions. Tap on "Cancel"

**Expected result**: The screen is dismissed
**Actual result**: Application crashes

### 2.
**Prerequisites**: No camera permissions
**Steps to reproduce**:
1. Follow steps above, but instead of "Take a picture", tap on "Record a video"
2. Test the same flow as above, by going to Settings and back without adding permission
3. Observe the black screen. Tap on it (or related actions)
4. The no permissions alert is displayed again. Tap on "Cancel" 

**Expected result**: The screen is dismissed
**Actual result**: Dialog is dismissed. Screen with black camera background still visible.

Basically both cases above describe the same flow.
In the first one, when taking a picture the issue is very apparent. The application should never crash.
The second one isn't so bad, since the user can still dismiss the screen by tapping the cancel button.
The second flow could be acceptable, however since we have to modify the code anyway, might as well do it right for both cases.

After testing out different cases, there seemed to be no other issues.
Resolving the ones above should mean the task can be closed from my side.
